### ISSUE TITLE
Please add Heraldicon integration

### ISSUE THREAD

**Issue by or** [mirnovov/heraldtron#1]
I've added an API now that should work for Heraldtron integration. I'll happily try to add code myself, but I haven't fully grokked how it works yet, so I thought I'd just document the API as a first step. And perhaps you have certain requirements that could be sorted out here as well.

## General
Site: https://heraldicon.org
API endpoint: https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api
Data format:
- I've added JSON support, to avoid the dependency below, all my examples will use JSON
  - required HTTP headers:
    - `Content-Type: application/json`
    - `Accept: application/json`
- natively I use [edn](https://github.com/edn-format/edn), which can be read and generated in Python via [edn-format](https://pypi.org/project/edn-format/)

### `generate-from-blazon`
POST Payload:
```json
{
  "call": "generate-from-blazon", 
  "data": {
    "blazon": "or, a lion azure langued gules"
  }
}
```
Success response:
```json
{
  "success": {
    "blazon": "or, a lion azure langued gules",
    "edit-url": "https://heraldicon.org/arms/new?base=905feefa44",
    "png-url": "https://data.heraldicon.org/blazonry-api/png/115905edf022a1c7fd4a1fa06937c6008198a817.png"
  }
}
```

Example:
```bash
curl -H "Content-Type: application/json" -H 'Accept: application/json' https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api --data '{"call": "generate-from-blazon", "data": {"blazon": "or, a lion azure langued gules"}}'
```

The `png-url` is the rendered arms and `edit-url` leads to the site to work on the arms with the parsed data already preloaded.

#### Error handling
If there's an error, then the response will have a HTTP status code of 400 (or 500 for internal server error) and the payload should look like this, e.g. for invalid JSON data:
```json
{
  "error": {
    "message": "Invalid JSON data.",
    "type":"client-invalid-json-data"
  }
}
```
If there's an error during the parsing of the blazon, then additional data is provided. In this example "langued" is spelled "langed", the `index` points to the beginning of the word that wasn't understood (`langed`) and `auto-complete -> choices` contains a list of `[suggestion, suggestion-type]` of words that the parser would've expected. The list is ordered by how well the suggestion fits and the type. In this example `langued` ends up on top. `error` usually will be `null`, unless there's an unexpected error... this probably can be ignored. I thought the auto-complete choices could be presented in some clever way as well, perhaps only the first few, if the list is too long and can't be hidden somehow.
```json
{
  "error": {
    "message": "Parser error",
    "type": "client-blazonry-parser-error",
    "data": {
      "value": "or, a lion azure langed gules",
      "error": null,
      "index": 17
      "auto-complete": {
        "choices": [
          ["langued", "extra tincture"],
          ["label", "ordinary"],
          ["labels", "ordinary"],
          ["leaved", "extra tincture"],
          ["legged", "extra tincture"],
          ["lion", "charge"],
          ["lions", "charge"],
          ["lozenge", "charge"],
          ["lozenges", "charge"],
          // ...
        ]
      }
    }
  }
}
```

#### Additional options
Additional options can be passed with a list of keywords under the key `options`:
```json
{
  "call": "generate-from-blazon",
  "data": {
    "blazon": "or, a lion azure langued gules",
    "options": [":swiss", ":squiggly", ":wikipedia-france"]
  }
}
```
These are just a bunch of special names, but they must be prefixed with a colon. Strings that are not recognized are simply ignored.
A list of supported options can be queried programmatically with the next API function, and I think that would be a useful feature for users as well.

### `blazon-options`
Payload:
```json
{
  "call": "blazon-options"
}
```
Response:
```json
{
  "success": {
    "options": {
      "escutcheon": [
        "english", "flag", "french-modern", "heater", "kite", "lozenge", "none", "norman",
        "norman-rounded", "oval", "polish", "polish-19th-century", "rectangle", "renaissance",
        "roundel", "square-czech", "square-french", "square-iberian", "swiss", "wedge"
      ],
      "mode": ["colours", "hatching"],
      "miscellaneous": ["outline", "shiny", "squiggly"],
      "theme": [
        "all", "community-cmwhyk", "community-content-cranium", "community-cotton-candy",
        "community-crystal-gems", "community-home-world", "community-jewelicious", "community-main-seven",
        "community-mother-earth", "community-pastell-puffs", "community-pretty-soldier", "community-rainbow-groom",
        "community-sodacan", "community-the-monet-maker", "community-van-goes-vroem", "ral-traffic",
        "wappenwiki", "web", "wikipedia-bajuvarian", "wikipedia-brandenburg", "wikipedia-default",
        "wikipedia-france", "wikipedia-hungary", "wikipedia-spain", "wikipedia-sweden", "wikipedia-switzerland",
        "wikipedia-web", "wikipedia-wurttemberg"
      ],
      "texture": [
        "cloth-rough", "cloth-smooth", "felt", "glass-frosted", "marble", "metal-brushed-1",
        "metal-brushed-2", "metal-rough", "metal-scraped", "metal-smooth", "mosaic", "none",
        "satin", "stone", "stone-rough", "wood"
      ]
    }
  }
}
```

**Comment by or** [comment]
I'm not sure what the best interface would be, perhaps you have some ideas.

I'm guessing some sort of trigger, e.g. "!hdn", perhaps an optional list of colon-prefixed options and the rest is blazon?
So like `!hdn :heater :outline per pale azure and argent, a pale wavy counterchanged` would be sent as:
```json
{
  "call": "generate-from-blazon",
  "data": {
    "blazon": "per pale azure and argent, a pale wavy counterchanged",
    "options": [":heater", ":outline"]
  }
}
```

**Comment by mirnovov** [comment]
Cool, thanks! If you don't get around to it, I'll add it myself in the coming months when I revamp the bot for slash commands.
I'm thinking `heraldicon` for the command as commands typically use long forms with short format as aliases (but as slash commands don't support aliases, I haven't decided what form I'll use there)

**Comment by mirnovov** [comment]
And yeah, the bot code is quite complex and makes use of many advanced Python features, as it's really designed for my use lol - let me know if you want any help.

**Comment by mirnovov** [comment]
Closed in 014f9c516f7d59a77b3c8488f0455724f78ee7a5
